connectors/mediasession/src/main/java/com/theoplayer/android/connector/mediasession/MediaSessionConnector.kt

```diff
@@ -0,0 +1,137 @@
+package com.theoplayer.android.connector.mediasession
+
+import android.media.session.PlaybackState.STATE_NONE
+import android.support.v4.media.MediaMetadataCompat
+import android.support.v4.media.session.MediaSessionCompat
+import android.util.Log
+import com.theoplayer.android.api.player.Player
+import com.theoplayer.android.api.source.SourceDescription
+
+const val TAG = "MediaSessionConnector"
+const val DEFAULT_MEDIA_SESSION_FLAGS = (MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS
+        or MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)
+const val EDITOR_MEDIA_SESSION_FLAGS =
+    DEFAULT_MEDIA_SESSION_FLAGS or MediaSessionCompat.FLAG_HANDLES_QUEUE_COMMANDS
+
+/**
+ * MediaSessionConnector connects an Android media session to a THEOplayer instance.
+ *
+ * <pre>
+ *   // Create and initialize the media session
+ *   mediaSession = MediaSessionCompat(this, TAG).apply {
+ *     // Do not let MediaButtons restart the player when the app is not visible
+ *     setMediaButtonReceiver(null)
+ *   }
+ *
+ *   // Create a MediaSessionConnector and attach the THEOplayer instance.
+ *   mediaSessionConnector = MediaSessionConnector(mediaSession)
+ *   mediaSessionConnector.player = player
+ *
+ *   // Optionally show debug logs
+ *   mediaSessionConnector.debug = true
+ *
+ *   // Set mediaSession to active
+ *   mediaSessionConnector.setActive(true)
+ * </pre>
+ *
+ * NOTE: The implementation is inspired by ExoPlayer's MediaSessionConnector.
+ *
+ * @see
+ *  <ul>
+ *      <li><a href="https://developer.android.com/guide/topics/media-apps/working-with-a-media-session">Using a media session</a></li>
+ *      <li><a href="https://developer.android.com/guide/topics/media-apps/video-app/building-a-video-app">Building a video app</a></li>
+ *      <li><a href="https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/ext/mediasession/MediaSessionConnector.html">ExoPlayer MediaSessionConnector</a></li>
+ *  </ul>
+ */
+class MediaSessionConnector(val mediaSession: MediaSessionCompat) {
+    private val mediaSessionCallback: MediaSessionCallback = MediaSessionCallback(this)
+    private val metadataProvider: MediaMetadataProvider
+    private val playbackStateProvider: PlaybackStateProvider = PlaybackStateProvider(this)
```

Comment:
Any reason we initialize `MediaSessionCallback` & `PlaybackStateProvider` here, meanwhile `MediaMetadataProvider` is initialized in the `init{}` block?